src/kaleidoscope/plugin/OneShot.cpp

```diff
@@ -22,266 +22,339 @@
-              Key_LeftControl.getFlags());
-  } else {
-    key = Key(LAYER_SHIFT_OFFSET + idx - 8,
-              KEY_FLAGS | SYNTHETIC | SWITCH_TO_KEYMAP);
-  }
+// These functions are particularly useful for ActiveModColor, which
+// could potentially use three different color values for the three
+// states (sticky | active && !sticky | pressed && !active).
 
-  handleKeyswitchEvent(key, UnknownKeyswitchLocation, key_state | INJECTED);
+// WARNING: There is no bounds checking done before getting the value
+// from the array. The caller must check `OneShot.isOneShotKey(key)`
+// before calling any of these functions.
+bool OneShot::isActive(Key key) {
+  uint8_t n = key.getRaw() - ranges::OS_FIRST;
+  return state_[n].active;
 }
 
-void OneShot::activateOneShot(uint8_t idx) {
-  injectNormalKey(idx, IS_PRESSED);
+bool OneShot::isSticky(Key key) {
+  uint8_t n = key.getRaw() - ranges::OS_FIRST;
+  return state_[n].sticky;
 }
 
-void OneShot::cancelOneShot(uint8_t idx) {
-  state_[idx].active = false;
-  injectNormalKey(idx, WAS_PRESSED);
+bool OneShot::isPressed(Key key) {
+  uint8_t n = key.getRaw() - ranges::OS_FIRST;
+  return state_[n].pressed;
 }
 
-EventHandlerResult OneShot::onKeyswitchEvent(Key &mapped_key, KeyAddr key_addr, uint8_t keyState) {
-  uint8_t idx = mapped_key.getRaw() - ranges::OS_FIRST;
+bool OneShot::isStickable(Key key) {
+  uint8_t n = key.getRaw() - ranges::OS_FIRST;
+  return state_[n].stickable;
+}
 
-  if (keyState & INJECTED)
-    return EventHandlerResult::OK;
+// ----------------------------------------------------------------------------
+// Other functions
 
-  if (!isActive()) {
-    if (!isOneShotKey_(mapped_key)) {
-      return EventHandlerResult::OK;
+// Cancel all active OneShot keys (if `cancel_stickies` is true) or
+// just non-sticky active OneShot keys. This function is called by
+// Escape-OneShot to release active OneShot keys.
+void OneShot::cancel(bool cancel_stickies) {
+  for (uint8_t i{0}; i < ONESHOT_KEY_COUNT; ++i) {
+    if (state_[i].active) {
+      if (state_[i].sticky && !cancel_stickies) {
+        continue;
+      }
+      releaseOneShot(i);
     }
+  }
+}
 
-    if (keyToggledOff(keyState)) {
-      state_[idx].pressed = false;
-    } else if (keyToggledOn(keyState)) {
-      start_time_ = Runtime.millisAtCycleStart();
-      state_[idx].position = key_addr.toInt();
-      state_[idx].pressed = true;
-      state_[idx].active = true;
-      prev_key_ = mapped_key;
+// ----------------------------------------------------------------------------
+// Vestigial functions?
 
-      activateOneShot(idx);
-    }
+// Is this function used by anything? It's part of the public API, but
+// none of the core plugins use it. Calling the event handler doesn't
+// seem like the right thing to do here, so perhaps the function
+// should be deprecated, then removed.
+void OneShot::inject(Key mapped_key, uint8_t key_state) {
+  onKeyswitchEvent(mapped_key, UnknownKeyswitchLocation, key_state);
+}
```

Comment:
I can't figure out a use case for this `inject()` function. None of the core plugins uses it, and I'm suspicious of calling the event handler like this, rather than calling `handleKeyswitchEvent()`.